flink-runtime/src/test/java/org/apache/flink/runtime/operators/sort/ReusingSortMergeOuterJoinIteratorITCase.java

```diff
@@ -0,0 +1,587 @@
+
+			final List<MutableObjectIterator<Record>> inList1 = new ArrayList<MutableObjectIterator<Record>>();
+			inList1.add(gen1Iter);
+			inList1.add(const1Iter);
+
+			final List<MutableObjectIterator<Record>> inList2 = new ArrayList<MutableObjectIterator<Record>>();
+			inList2.add(gen2Iter);
+			inList2.add(const2Iter);
+
+			MutableObjectIterator<Record> input1 = new MergeIterator<Record>(inList1, comparator1.duplicate());
+			MutableObjectIterator<Record> input2 = new MergeIterator<Record>(inList2, comparator2.duplicate());
+
+			// collect expected data
+			final Map<TestData.Key, Collection<Match>> expectedMatchesMap = joinValues(
+					collectData(input1),
+					collectData(input2),
+					outerJoinType);
+
+			// re-create the whole thing for actual processing
+
+			// reset the generators and iterators
+			generator1.reset();
+			generator2.reset();
+			const1Iter.reset();
+			const2Iter.reset();
+			gen1Iter.reset();
+			gen2Iter.reset();
+
+			inList1.clear();
+			inList1.add(gen1Iter);
+			inList1.add(const1Iter);
+
+			inList2.clear();
+			inList2.add(gen2Iter);
+			inList2.add(const2Iter);
+
+			input1 = new MergeIterator<Record>(inList1, comparator1.duplicate());
+			input2 = new MergeIterator<Record>(inList2, comparator2.duplicate());
+
+			final FlatJoinFunction matcher = new MatchRemovingMatcher(outerJoinType, expectedMatchesMap);
+
+			final Collector<Record> collector = new DiscardingOutputCollector<Record>();
+
+
+			// we create this sort-merge iterator with little memory for the block-nested-loops fall-back to make sure it
+			// needs to spill for the duplicate keys
+			ReusingMergeOuterJoinIterator<Record, Record, Record> iterator =
+					new ReusingMergeOuterJoinIterator<Record, Record, Record>(
+							outerJoinType, input1, input2, serializer1, comparator1, serializer2, comparator2,
+							pairComparator, this.memoryManager, this.ioManager, PAGES_FOR_BNLJN, this.parentTask);
+
+			iterator.open();
+
+			while (iterator.callWithNextKey(matcher, collector)) ;
+
+			iterator.close();
+
+			// assert that each expected match was seen
+			for (Entry<TestData.Key, Collection<Match>> entry : expectedMatchesMap.entrySet()) {
+				if (!entry.getValue().isEmpty()) {
+					Assert.fail("Collection for key " + entry.getKey() + " is not empty");
+				}
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+			Assert.fail("An exception occurred during the test: " + e.getMessage());
+		}
+	}
+
+
+
+	// --------------------------------------------------------------------------------------------
+	//                                    Utilities
+	// --------------------------------------------------------------------------------------------
+
+
+	/**
+	 * custom function that simply joins two tuples and considers null cases
+	 */
+	private static class SimpleFlatJoinFunction implements FlatJoinFunction<Tuple2<String, String>, Tuple2<String, Integer>, Tuple4<String, String, String, Object>> {
```

Comment:
The utilities of the `ReusingSortMergeOuterJoinIteratorITCase` and `NonReusingSortMergeOuterJoinIteratorITCase` look very similar (didn't check in detail). Can you put them in an abstract class that both test classes extend?